File: modules/expo-ssl-trust/ios/ExpoSslTrustModule.swift
import ExpoModulesCore
import Foundation
import Security

public class ExpoSslTrustModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoSslTrust")

        // At app launch (before RN makes its first request and caches its
        // NSURLSession): load the trust store and install the URLSession swizzle
        // so RN fetch / images / downloads consult SslTrustURLProtocol for
        // trusted self-signed hosts — the Android-equivalent "just works" path.
        // The local proxy is used only for AVPlayer streaming.
        OnCreate {
            SslTrustStore.shared.initialize()
            URLSessionConfiguration.sslTrustInstallSwizzle()
        }

        AsyncFunction("initTrustStore") { (promise: Promise) in
            SslTrustStore.shared.initialize()
            // Resolve the install status to match the Android contract
            // (`{ installed, error }`). iOS has no JSSE-style install failure,
            // so `error` is always null once initialize() has run.
            promise.resolve([
                "installed": SslTrustStore.shared.installSucceeded,
                "error": NSNull(),
            ])
        }

        AsyncFunction("getInstallStatus") { (promise: Promise) in
            promise.resolve([
                "installed": SslTrustStore.shared.installSucceeded,
                "error": NSNull(),
            ])
        }

        AsyncFunction("getCertificateInfo") { (url: String, promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                do {
                    let info = try CertificateInspector.getCertificateInfo(urlString: url)
                    
                    // Store the DER data temporarily keyed by hostname
                    // so it's available when the user trusts the certificate
                    if let derData = info.derData,
                       let urlObj = URL(string: url),
                       let host = urlObj.host {
                        SslTrustStore.shared.storeCertificateData(hostname: host, derData: derData)
                    }
                    
                    let result: [String: Any] = [
                        "subject": info.subject,
                        "issuer": info.issuer,
                        "sha256Fingerprint": info.sha256Fingerprint,
                        "validFrom": info.validFrom,
                        "validTo": info.validTo,
                        "serialNumber": info.serialNumber,
                        "isSelfSigned": info.isSelfSigned
                    ]
                    promise.resolve(result)
                } catch {
                    promise.reject("ERR_CERT_FETCH", "Failed to fetch certificate: \(error.localizedDescription)")
                }
            }
        }
        
        AsyncFunction("trustCertificate") { (hostname: String, sha256Fingerprint: String, validTo: String?, promise: Promise) in
            SslTrustStore.shared.initialize()
            SslTrustStore.shared.trustCertificate(hostname: hostname, sha256Fingerprint: sha256Fingerprint, validTo: validTo)
            promise.resolve(nil)
        }

        AsyncFunction("removeTrustedCertificate") { (hostname: String, promise: Promise) in
            SslTrustStore.shared.initialize()
            SslTrustStore.shared.removeTrustedCertificate(hostname: hostname)
            promise.resolve(nil)
        }

        AsyncFunction("clearAllTrustedCertificates") { (promise: Promise) in
            SslTrustStore.shared.initialize()
            SslTrustStore.shared.clearAllTrustedCertificates()
            promise.resolve(nil)
        }
        
        AsyncFunction("getTrustedCertificates") { (promise: Promise) in
            SslTrustStore.shared.initialize()
            let certs = SslTrustStore.shared.getTrustedCertificates()
            promise.resolve(certs)
        }
        
        AsyncFunction("isCertificateTrusted") { (hostname: String, promise: Promise) in
            SslTrustStore.shared.initialize()
            let trusted = SslTrustStore.shared.isCertificateTrusted(hostname: hostname)
            promise.resolve(trusted)
        }

        // Reconcile the local reverse proxy's upstreams with the app's
        // configured server base URLs. Only URLs whose host has a trusted
        // self-signed cert are proxied; everything else connects directly.
        // Returns `{ port, upstreams: [{ baseUrl, token }] }` or null when no
        // upstream needs proxying. iOS only — Android has no proxy (its
        // OkHttp/TrustManager path handles trust at the HTTP layer).
        AsyncFunction("syncProxyUpstreams") { (baseUrls: [String], promise: Promise) in
            SslTrustStore.shared.initialize()
            let trusted = baseUrls.filter { url in
                guard let host = URL(string: url)?.host else { return false }
                return SslTrustStore.shared.isCertificateTrusted(hostname: host)
            }
            for url in trusted { _ = SslTrustProxy.shared.register(baseUrl: url) }
            SslTrustProxy.shared.retainOnly(trusted)
            if trusted.isEmpty {
                SslTrustProxy.shared.stop()
                promise.resolve(NSNull())
            } else {
                promise.resolve(SslTrustProxy.shared.ensureRunningAndWait() ?? NSNull())
            }
        }

        // Current proxy info (live port + token map). Restarts the listener if
        // it dropped (e.g. after a background suspend), so the JS cache can be
        // refreshed on foreground. Null when nothing is registered.
        AsyncFunction("getProxyInfo") { (promise: Promise) in
            promise.resolve(SslTrustProxy.shared.ensureRunningAndWait() ?? NSNull())
        }
    }
}

File: modules/expo-ssl-trust/android/src/main/java/expo/modules/ssltrust/ExpoSslTrustModule.kt
package expo.modules.ssltrust

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class ExpoSslTrustModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoSslTrust")

        AsyncFunction("initTrustStore") { promise: Promise ->
            try {
                val context = appContext.reactContext ?: throw Exception("React context not available")
                SslTrustStore.init(context)
                // SslTrustStore.init() catches its own JSSE failures and records them
                // on installSucceeded / lastInstallError. The JS layer queries
                // getInstallStatus() to surface failures to the user.
                promise.resolve(
                    mapOf(
                        "installed" to SslTrustStore.installSucceeded,
                        "error" to SslTrustStore.lastInstallError,
                    )
                )
            } catch (e: Exception) {
                promise.reject("ERR_INIT_TRUST_STORE", e.message, e)
            }
        }

        AsyncFunction("getInstallStatus") { promise: Promise ->
            promise.resolve(
                mapOf(
                    "installed" to SslTrustStore.installSucceeded,
                    "error" to SslTrustStore.lastInstallError,
                )
            )
        }

        AsyncFunction("getCertificateInfo") { url: String, promise: Promise ->
            CoroutineScope(Dispatchers.IO).launch {
                try {
                    val info = CertificateInspector.getCertificateInfo(url)
                    
                    // Store the DER data temporarily so it's available
                    // when the user trusts the certificate
                    if (info.derData != null) {
                        try {
                            val parsedUrl = java.net.URL(
                                if (url.startsWith("http")) url else "https://$url"
                            )
                            SslTrustStore.storeCertDerData(parsedUrl.host, info.derData)
                        } catch (_: Exception) { }
                    }
                    
                    val result = mapOf(
                        "subject" to info.subject,
                        "issuer" to info.issuer,
                        "sha256Fingerprint" to info.sha256Fingerprint,
                        "validFrom" to info.validFrom,
                        "validTo" to info.validTo,
                        "serialNumber" to info.serialNumber,
                        "isSelfSigned" to info.isSelfSigned
                    )
                    promise.resolve(result)
                } catch (e: Exception) {
                    promise.reject("ERR_CERT_FETCH", "Failed to fetch certificate: ${e.message}", e)
                }
            }
        }

        AsyncFunction("trustCertificate") { hostname: String, sha256Fingerprint: String, validTo: String?, promise: Promise ->
            try {
                val context = appContext.reactContext ?: throw Exception("React context not available")
                SslTrustStore.init(context) // trusting needs the trust-manager install
                SslTrustStore.trustCertificate(hostname, sha256Fingerprint, validTo)
                promise.resolve(null)
            } catch (e: Exception) {
                promise.reject("ERR_TRUST_CERT", e.message, e)
            }
        }

        AsyncFunction("removeTrustedCertificate") { hostname: String, promise: Promise ->
            try {
                val context = appContext.reactContext ?: throw Exception("React context not available")
                // loadCerts (not init): removeTrustedCertificate re-installs the
                // trust manager itself with the updated store.
                SslTrustStore.loadCerts(context)
                SslTrustStore.removeTrustedCertificate(hostname)
                promise.resolve(null)
            } catch (e: Exception) {
                promise.reject("ERR_REMOVE_CERT", e.message, e)
            }
        }

        AsyncFunction("clearAllTrustedCertificates") { promise: Promise ->
            try {
                val context = appContext.reactContext ?: throw Exception("React context not available")
                SslTrustStore.loadCerts(context)
                SslTrustStore.clearAllTrustedCertificates()
                promise.resolve(null)
            } catch (e: Exception) {
                promise.reject("ERR_CLEAR_CERTS", e.message, e)
            }
        }

        // Read-only: `loadCerts` (no install). The OkHttp factory is installed at
        // app startup by ExpoSslTrustPackage's ApplicationLifecycleListener (which
        // is the only point early enough — before RN builds its fixed HTTP client),
        // so reads don't need to install.
        AsyncFunction("getTrustedCertificates") { promise: Promise ->
            try {
                val context = appContext.reactContext ?: throw Exception("React context not available")
                SslTrustStore.loadCerts(context)
                val certs = SslTrustStore.getTrustedCertificates()
                promise.resolve(certs)
            } catch (e: Exception) {
                promise.reject("ERR_GET_CERTS", e.message, e)
            }
        }

        AsyncFunction("isCertificateTrusted") { hostname: String, promise: Promise ->
            try {
                val context = appContext.reactContext ?: throw Exception("React context not available")
                // loadCerts only: not called from JS in a path that needs install
                // (the native hostname verifier reads the store directly).
                SslTrustStore.loadCerts(context)
                val trusted = SslTrustStore.isCertificateTrusted(hostname)
                promise.resolve(trusted)
            } catch (e: Exception) {
                promise.reject("ERR_CHECK_CERT", e.message, e)
            }
        }
    }
}
